jcl/src/java.base/share/classes/java/lang/String.java

```diff
@@ -3891,61 +3891,136 @@ static void checkBoundsBeginEnd(int begin, int end, int length) {
+			private int cursor; // next character to read
 
-		@Override
-		public int characteristics() {
-			return IMMUTABLE | ORDERED | SIZED | SUBSIZED;
-		}
+			CharSpliterator(int start, int end) {
+				this.start = start;
+				this.end = end;
+				this.cursor = start;
+			}
+
+			@Override
+			public long estimateSize() {
+				return end - start;
+			}
 
-		@Override
-		public OfInt trySplit() {
-			int newEnd = cursor + (end - cursor + 1) / 2;
-			StringSpliteratorOfInt newSpliterator = null;
-			if (newEnd < end) {
-				newSpliterator = new StringSpliteratorOfInt(newEnd, end, func);
+			@Override
+			public int characteristics() {
+				return IMMUTABLE | ORDERED | SIZED | SUBSIZED;
+			}
+
+			@Override
+			public OfInt trySplit() {
+				int oldEnd = end;
+				int newEnd = cursor + (oldEnd - cursor + 1) / 2;
+				if (newEnd >= oldEnd) {
+					return null;
+				}
 				end = newEnd;
+				return new CharSpliterator(newEnd, oldEnd);
 			}
-			return newSpliterator;
-		}
 
-		@Override
-		public boolean tryAdvance(IntConsumer action) {
-			boolean result = false;
-			if (cursor < end) {
-				action.accept(func.applyAsInt(cursor));
-				result = true;
-				++cursor;
+			@Override
+			public boolean tryAdvance(IntConsumer action) {
+				if (cursor >= end) {
+					return false;
+				}
+				action.accept(charAtInternal(cursor));
+				cursor += 1;
+				return true;
 			}
-			return result;
 		}
+
+		return StreamSupport.intStream(new CharSpliterator(0, lengthInternal()), false);
 	}
-	
-	@Override
-	public IntStream chars() {
-		return StreamSupport.intStream(new StringSpliteratorOfInt(0, lengthInternal(), i -> charAt(i)), false);
-	}
-	
+
 	@Override
 	public IntStream codePoints() {
-		return StreamSupport.intStream(new StringSpliteratorOfInt(0, lengthInternal(), i -> codePointAt(i)), false);
+		final class CodeSpliterator implements Spliterator.OfInt {
+			private final int start; // character index of the first code point
+			private int end; // character index after the final code point
+			private int cursor; // character index of the next code point
+
+			CodeSpliterator(int start, int end) {
+				this.start = start;
+				this.end = end;
+				this.cursor = start;
+			}
+
+			@Override
+			public long estimateSize() {
+				return codePointCount(start, end);
+			}
+
+			@Override
+			public int characteristics() {
+				return IMMUTABLE | ORDERED | SIZED | SUBSIZED;
+			}
+
+			@Override
+			public OfInt trySplit() {
+				int oldEnd = end;
+				int newEnd = cursor + (oldEnd - cursor + 1) / 2;
+
+				if (newEnd >= oldEnd) {
+					return null;
+				}
+
+				/*
+				 * If newEnd is the character index of a low surrogate, we include
+				 * the containing code point in this spliterator and only if there
+				 * are code points after it do we return a new spliterator.
+				 */
+				if (newEnd > start
+						&& Character.isHighSurrogate(charAtInternal(newEnd - 1))
+						&& Character.isLowSurrogate(charAtInternal(newEnd))) {
```

Comment:
Do we need both tests?  If the `char[newEnd - 1]` is a high surrogate, then splitting in the middle of the pair seems wrong

Reply:
The condition is consistent with treatment of code points elsewhere. See `codePointAt()` for example where a high surrogate must be followed by a low surrogate to be treated as a pair.